### ISSUE TITLE
[themes][semantic] default themes: new constant color

### ISSUE THREAD

**Issue by aeschli** [microsoft/vscode#95346]
Dark+ / Light +
Use the default color for constant variables.

- as done in the VS 2019: See #92965
- take advantage of constants info from semantic highlighting

**Comment by mjbvz** [comment]
Hey @aeschli just saw this in today's insiders.

For JS/TS, the new color for constants does change the character of the dark theme. One specific nit I have is that constants now use the same color as punctuation which makes code less legible:

![Screen Shot 2020-04-16 at 2 39 08 PM](https://user-images.githubusercontent.com/12821956/79509288-120de300-7ff0-11ea-90fa-2d6dfa63422b.png)

Trying to match VS is a good goal but I find the impact of this little change to be quite drastic (more drastic than enabling semantic highlighting in the first place). And honestly, we have a whole lot of users who are using the default VS Code theme and don't care what Visual Studio is doing

Maybe there's a compromise that won't be quite as jarring when users upgrade but still lets us convey if something is a constant or not?

**Comment by JonWallsten** [comment]
I thought it was a bug and was about to report it. It feels like the color that should be there just disappeared.

**Comment by aeschli** [comment]
I fully understand. It also took me a while to like it. Theme changes are always very controversial.
Still, I wanted to put it out to see the general reaction. I'm fine to revert it for stable if there are many negative reactions.
- I first tried to find another color, but it's hard and equally controversial. Also there has been many requests for more colors in the default themes and we always resisted with the argument to keep it simple and clean
- we define a new Default theme that we only give as default to new users. If we do that, it should be more than just a single color change, but a a new, fresh design.
@misolori Your thoughts?

**Comment by JonWallsten** [comment]
@aeschli: Would it be possible to use something in between cyan and white? Or go darker. 
The reason I got the feeling something was wrong with the theme was because nothing else in the dark theme is white. Everything is different colors. So it felt like something was "missing". Personally I don't care if let/const have the same color, but since they are still very similar from a semantic POV maybe a more subtle color change would suffice?

**Comment by IllusionMH** [comment]
> The change of color for consts in a test run in Insiders. Please give it a chance and please add a thumbs up/down on that issue.
From https://github.com/microsoft/vscode/issues/95434#issuecomment-615086755 

It's a bit strange to close opened issue as duplicate of this and expect feedback on closed issue. Could you have an open one for feedback?

I don't have any preference for color, however default "white" doesn't have any as with plain text and would be nice to have something else.

---

> take advantage of constants info from semantic highlighting

I don't think that knowledge that some variable/property is constant is more useful to discard type of data in variable as was implemented for JS/TS.
In JS some teams (our too) use `const` for everything except that we need reassignments `let` (counts are `const` 10k+ vs `let` ~1.5k) so almost every variable in our app is white.
It's also (almost ?) impossible to create `readonly` object properties in plain JS - no changes here.

Previous behavior with value type should have greater priority than generic constant.

**Comment by aeschli** [comment]
Reopened for feedback

**Comment by mattacosta** [comment]
I'm not finding this particularly useful, especially on properties. You have to know something about the kind of immutability being used in order to determine if/when a property can be written to.

Unfortunately, some extensions/grammars do not respect the subtle distinctions. For example, in TypeScript, while the `readonly` modifier is correct semantically, it's not the same as a `const` modifier, and yet that is the scope being used on read-only properties. If you switch to a language with properly scoped read-only fields, then you now have two colors for the same thing.

This is also made worse if a language server is inconsistent in its semantic highlighting. Again, from TypeScript:

```ts
class A {

  public readonly roField = 1;

  // Without a setter, this should be read-only too, especially since it
  // could also implement an abstract readonly field from an interface.
  public get roProperty(): number { return 1; }

  public foo(): number {
    return this.roField /* white */ + this.roProperty /* light blue */;
  }

}
```

Another problem is that enum members aren't easily distinguishable in code anymore. This is mainly because the light green color was replaced and isn't being used on them or the enum identifier. It would be a lot better if the enum identifier got that color by default now, which would also be consistent with VS2019 as well.

That said, I would like to see the default theme be a little closer to VS2019. Another alternative could be to reduce the change to just *local* constants (if possible, pending other issues) and enum members (which are never ambiguous because all members are always `const`).

**Comment by MirkSerain** [comment]
I also thought this was a bug when I saw it.  I agree with @IllusionMH, my JS team also uses const for everything and now I have a lot of white text in my dark theme.  It just feels like part of the highlighting isn't working or that the variable i'm using doesn't exist.

**Comment by aeschli** [comment]
We will give an new color a try that is closer to the variable color. Please try it out a while and let me know whether the new color is better or if we should go back to the original theme.

![image](https://user-images.githubusercontent.com/6461412/80073029-601e6b80-8547-11ea-8cbf-f85237c75658.png)

**Comment by tamuratak** [comment]
Version: 1.45.0-insider
Commit: a250df703de955a38aed427a917bce8278ab3331
Date: 2020-04-24T20:46:59.782Z
Electron: 7.2.2
Chrome: 78.0.3904.130
Node.js: 12.8.1
V8: 7.8.279.23-electron.0
OS: Darwin x64 17.7.0

I am a user of `Default Light+`. The current  color of constant variables,  `variable.other.constant,variable.other.enummember { "foreground": "#328267" }`, is too light when the background color is white. 

It is difficult to distinguish it from the color of comments.

1.44.2:

![スクリーンショット 2020-04-25 22 25 02](https://user-images.githubusercontent.com/10665499/80281179-54e45f00-8744-11ea-9edd-391258023fd5.png)

Insiders:

![スクリーンショット 2020-04-25 22 25 11](https://user-images.githubusercontent.com/10665499/80281183-5f9ef400-8744-11ea-96fc-60c1b6ad93e1.png)

**Comment by mattacosta** [comment]
There's not a lot of contrast with existing colors in the dark theme either. It's kind of in no man's land between `this`, type names, and unused variables/parameters.

![const_color](https://user-images.githubusercontent.com/1007463/80228666-2cede200-8604-11ea-8816-250a0b0abb2f.png)

**Comment by aeschli** [comment]
@misolori see above, comment on the contrast of the new color.

**Comment by misolori** [comment]
@tamuratak thanks for the feedback, our accessibility color contrast goals are to meet WCAG AA so you will find this throughout all of our themes. The new const color [meets WCAG AA at 4.6:1](https://webaim.org/resources/contrastchecker/?fcolor=328267&bcolor=FFFFFF), see below.

![image](https://user-images.githubusercontent.com/35271042/80402355-a854cb00-8872-11ea-9a6a-a4387cffef51.png)


cc @isidorn

**Comment by aeschli** [comment]
To verify:
Open the Dark+/Light+ themes on a TypeScript file and look at the coloring of const variables, enum members and read-only properties . They should have a different color than regular variables.
Verify that the color is not to be confused from the keyword and commend color.

**Comment by LeeviHalme** [comment]
Can you please remove it? Or at least make it toggleable. I thought this was a bug because it was so ugly lol

**Comment by resetus** [comment]
Default Dark Theme:  

- x - variable is not used
- y - variable is used
- a - variable is used
- b - variable is not used

![image](https://user-images.githubusercontent.com/24482422/81401257-9ce98580-9137-11ea-93c4-aa498546e16b.png)

This color creates more problems than benefits

**Comment by chybisov** [comment]
I loved my yellow enums! Bring them back! 😢

**Comment by gnaeus** [comment]
Declaration and usage of the same variable have different colors. Looks a bit strange:

![image](https://user-images.githubusercontent.com/4706707/81417905-93224b00-9154-11ea-90a7-4d99f4bdd4f9.png)

**Comment by aeschli** [comment]
@gnaeus Did you disable semantic highlighting?

**Comment by resetus** [comment]
@aeschli, why **console.log** changes color, when this option is enabled?

![image](https://user-images.githubusercontent.com/24482422/81419541-16dd3700-9157-11ea-83f9-93461217942b.png)

**Comment by gnaeus** [comment]
@aeschli no I didn't disable semantic highlighting. My combination of flags:
```json
    "editor.semanticHighlighting.enabled": true,
    "editor.semanticTokenColorCustomizations": {
        "enabled": false,
    },
```

**Comment by gnaeus** [comment]
Also here is the settings for all who loves __the old behavior__:

```jsonc
"editor.semanticTokenColorCustomizations": {
  "rules": {
    // different color for all constants
    "*.readonly": "#9cdcfe",
  }
}
```

<table>
<tr><th>OLD</th><th>NEW</th></tr>
<tr>
<td><img width="300" src="https://user-images.githubusercontent.com/4706707/81421790-b819bc80-915a-11ea-9a2b-e004b3477316.png" /></td>
<td><img width="300" src="https://user-images.githubusercontent.com/4706707/81421594-5b1e0680-915a-11ea-859a-f51ce6da1348.png" /></td>
</tr>
</table>

**Comment by chybisov** [comment]
@gnaeus enums should be yellow :)

**Comment by sashahavia** [comment]
Can someone return back the original color for constants? It's clashing with the green color. Not sure why this was done

I tried using this to return to old behavior and it doesn't work

```
"editor.semanticTokenColorCustomizations": {
  "rules": {
    // different color for all constants
    "*.readonly": "#9cdcfe",
  }
}
```

**Comment by aeschli** [comment]
To get the old colors back you need to add the following to the settings:
```
    "editor.tokenColorCustomizations": {
        "[Default Dark+]": {
            "textMateRules": [
                {
                    "scope": ["variable.other.constant", "variable.other.enummember"],
                    "settings": {
                        "foreground": "#9CDCFE"
                    }
                }
            ]
        },
        "[Default Light+]": {
            "textMateRules": [
                {
                    "scope": ["variable.other.constant", "variable.other.enummember"],
                    "settings": {
                        "foreground": "#001080"
                    }
                }
            ]
        },
    },
```

**Comment by sashahavia** [comment]
> To get the old colors back you need to add the following to the settings:
> 
> ```
>     "editor.tokenColorCustomizations": {
>         "[Default Dark+]": {
>             "textMateRules": [
>                 {
>                     "scope": ["variable.other.constant", "variable.other.enummember"],
>                     "settings": {
>                         "foreground": "#9CDCFE"
>                     }
>                 }
>             ]
>         },
>         "[Default Light+]": {
>             "textMateRules": [
>                 {
>                     "scope": ["variable.other.constant", "variable.other.enummember"],
>                     "settings": {
>                         "foreground": "#001080"
>                     }
>                 }
>             ]
>         },
>     },
> ```

Thank you for a quick response. It worked

**Comment by arielkuzminski** [comment]
Just spend 30 minutes trying to fix this bug by reopening VSC then reinstalling it because I was sure that my coloring is broken.
Now I read that, this is a "feature".

**Comment by LeeviHalme** [comment]
> To get the old colors back you need to add the following to the settings:
> 
> ```
>     "editor.tokenColorCustomizations": {
>         "[Default Dark+]": {
>             "textMateRules": [
>                 {
>                     "scope": ["variable.other.constant", "variable.other.enummember"],
>                     "settings": {
>                         "foreground": "#9CDCFE"
>                     }
>                 }
>             ]
>         },
>         "[Default Light+]": {
>             "textMateRules": [
>                 {
>                     "scope": ["variable.other.constant", "variable.other.enummember"],
>                     "settings": {
>                         "foreground": "#001080"
>                     }
>                 }
>             ]
>         },
>     },
> ```

Thank you! It is unbelievable that WE as users have to make our own fixes around this new "feature". Though you guys at Microsoft would listen to us... :unamused:

**Comment by evg656e** [comment]
Holy moly, I have color blindness, and after recent update, I thought it some kind of bug. You should  change color schemes with care. Reverting to old style

**Comment by arielkuzminski** [comment]
> Holy moly, I have color blindness, and after recent update, I thought it some kind of bug. You should change color schemes with care. Reverting to old style

Exactly. I'm color blind too, so that's why I have major problem with this. I guess this feature hasn't been consulted with UX and hasn't been tested.
Second major bad feature for me in VSC in 2020. I wonder if it's beggining of downfall of VSC and some new "light, fresh" player should appear on market.

**Comment by luisdiegohernandez** [comment]
Leaving a comment here just for feedback. I feel a little too picky about this, but I spent more than an hour trying to figure out what was wrong with the const coloring. I thought it was a bug or some intrusive extension messing up my settings. Thanks for the "solution"!

**Comment by vnestere** [comment]
Spent couple hours today reinstalling vscode, was sure its a bug. Had to go with color customization fix mentioned above to revert it back to normal colors.

Why would you add such "features" that break normal workflow for so many people? I'm wondering how many hours collectively developers who use vscode have wasted trying to work it around back to normal.

**Comment by mattacosta** [comment]
See #97373 for the discussion about improving the color.

**Comment by giuliano-december** [comment]
> > To get the old colors back you need to add the following to the settings:
> > ```
> >     "editor.tokenColorCustomizations": {
> >         "[Default Dark+]": {
> >             "textMateRules": [
> >                 {
> >                     "scope": ["variable.other.constant", "variable.other.enummember"],
> >                     "settings": {
> >                         "foreground": "#9CDCFE"
> >                     }
> >                 }
> >             ]
> >         },
> >         "[Default Light+]": {
> >             "textMateRules": [
> >                 {
> >                     "scope": ["variable.other.constant", "variable.other.enummember"],
> >                     "settings": {
> >                         "foreground": "#001080"
> >                     }
> >                 }
> >             ]
> >         },
> >     },
> > ```
> 
> Thank you! It is unbelievable that WE as users have to make our own fixes around this new "feature". Though you guys at Microsoft would listen to us... 😒

This is awesome. I love you man. I was going nuts with the new coloring

**Comment by oskarkook** [comment]
The issues with this:
1. It provides very little value. I don't really care whether a variable is a const or not. In the cases where I do, I will get feedback that it's a const very quickly anyway.
2. It's hard to parse and is confusing. By making every little thing have a different color, it ends up being very hard to remember that color X stands for Y thing. As a result, it just confuses you because you're looking at rainbow-colored text. Different colors should stand for very significant differences where it makes it _easier_ to read text than _harder_. (See how I used italics for specific words to _enhance_ my point? If the whole text was a mixture of bold and italics, it would be very hard to read.)
3. The color does not fit into the theme. The contrast against the dark background is not good enough in my opinion, even though you claim that it satisfies some standard. The usual (mutable) variable color works really well against the dark background, I prefer it highly over this color. Most of my variables are constants, so this is a serious eyesore.
